services/user/Registry/service/src/lib.rs

```diff
@@ -0,0 +1,378 @@
+            existing_tag.clone()
+        } else {
+            create_new_app_tag(tag, account_id)
+        }
+    }
+
+    fn create_new_app_tag(tag: &str, account_id: AccountNumber) -> AppTags {
+        let app_tags_table = AppTagsTable::new();
+        let tag_id = get_or_create_tag_id(tag);
+        let new_app_tag = AppTags {
+            app_id: account_id,
+            tag_id,
+        };
+        app_tags_table.put(&new_app_tag).unwrap();
+        new_app_tag
+    }
+
+    fn find_existing_app_tag(tag: &str, existing_app_tags: &Vec<AppTags>) -> Option<AppTags> {
+        let tags_table = TagsTable::new();
+        existing_app_tags
+            .iter()
+            .find(|at| {
+                tags_table
+                    .get_index_pk()
+                    .get(&at.tag_id)
+                    .map(|tr| tr.tag == tag)
+                    .unwrap_or(false)
+            })
+            .cloned()
+    }
+
+    fn updateAppTags(account_id: AccountNumber, tags: &Vec<String>) {
+        let app_tags_table = AppTagsTable::new();
+
+        // Get existing app tags
+        let existing_app_tags: Vec<AppTags> = app_tags_table
+            .get_index_pk()
+            .range((account_id, 0)..(account_id, u32::MAX))
+            .collect();
+
+        let mut new_app_tags = Vec::new();
+
+        for tag in tags {
+            let app_tag = process_tag(tag, account_id, &existing_app_tags);
+            new_app_tags.push(app_tag);
+        }
+
+        remove_obsolete_tags(&existing_app_tags, &new_app_tags);
+
+        // todo: validate if tags are valid (lowercase alphanumeric and dashes)
+    }
+
+    fn get_or_create_tag_id(tag: &str) -> u32 {
+        let tags_table = TagsTable::new();
+        let idx = tags_table.get_index_by_tags();
+        idx.range((tag.to_string(), 0)..(tag.to_string(), u32::MAX))
+            .find(|tag_row| tag_row.tag == tag)
+            .map(|tag_row| tag_row.id)
+            .unwrap_or_else(|| create_new_tag(tag.to_string()))
+    }
+
+    fn create_new_tag(tag: String) -> u32 {
+        let tags_table = TagsTable::new();
+        let new_id = tags_table
+            .get_index_pk()
+            .into_iter()
+            .last()
+            .map(|last_tag| last_tag.id + 1)
+            .unwrap_or(1);
+        let new_tag = TagRecord {
+            id: new_id,
+            tag: tag.to_string(),
+        };
+        tags_table.put(&new_tag).unwrap();
+        new_id
+    }
+
+    fn remove_obsolete_tags(existing_app_tags: &[AppTags], new_app_tags: &[AppTags]) {
+        let app_tags_table = AppTagsTable::new();
+        for existing_tag in existing_app_tags {
+            if !new_app_tags
+                .iter()
+                .any(|at| at.tag_id == existing_tag.tag_id)
+            {
+                app_tags_table.remove(existing_tag);
+                // TODO: Check if there are any apps using the tag from tags_table and remove if not
```

Comment:
Also a good idea.

Can a range query hold the second tuple element constant and modulate the first? If so then I think it should be as easy as:

```
if app_tags_table
    .get_index_pk()
    .range((0, old_tag.tag_id)..(u64::MAX, old_tag.tag_id))
    .count() == 0 {
        tags_table.remove(old_tag.tag_id);
    }
```

Actually, you probably want a secondary index for app_tags_table that reverses the tuple order anyway, to make filtering by tag more efficient.